docs_src/design/adr/application/0025-Record-and-Replay.md

```diff
@@ -0,0 +1,128 @@
+- New pipeline function written to process the batched data once the batching threshold has been exceeded. This function will simply send the recorded data to an async function for processing. 
+
+The async  function receiving the data will first [stop the Functions Pipeline](https://docs.edgexfoundry.org/2.3/microservices/application/ApplicationServiceAPI/#makeitstop) and then save the data for later replay and/or export. It will also determine the list of unique Device Profile and Device Names from the data and store them along side the recorded data. Since app services can receive Events out of order per their timestamps, the saved Event data must be sorted by the Event timestamps. All data will saved in in-memory storage. 
+
+!!! note
+    Starting a new recording will overwrite any previous recorded data.
+
+#### Record Request DTO
+
+##### Duration
+
+Time duration in which to record data. Required if **Event Limit** is not specified.
+
+##### Event Limit 
+
+Maximum number `Events` to record. Required if **Duration** is not specified
+
+##### Include Device Profile Names 
+
+Optional list of Device Profile Names to filter for
+
+##### Include Device Names
+
+Optional list of Device Names to filter for
+
+##### Exclude Device Profile Names
+
+Optional list of Device Profile Names to filter out
+
+##### Exclude Device Names
+
+Optional list of Device Names to filter out
+
+### Replay endpoint
+
+This `POST` API will start replaying the recorded data as specified in the [request Data Transfer Object (DTO)](#replay-request-dto) defined below. The request handler will validate the DTO and that the appropriate Device Profiles and Devices from the data exist. It will then start an async Go function to handle the replay so the request doesn't timeout on long replays. 
+
+The replay async Go function will use the [Background Publishing](https://docs.edgexfoundry.org/2.3/microservices/application/AdvancedTopics/#background-publishing) capability to send the recorded Events to the EdgeX MessageBus using the same publish topic scheme used by Device Services, which is `edgex/events/device/<device-profile-name>/<device-name>/<source-name>`. The App SDK has the  [Publish Topic Placeholders](https://docs.edgexfoundry.org/2.3/microservices/application/Triggers/#publish-topic-placeholders) capability built-in to facilitate this. The data for these topics is available from the Event DTO. The timestamps in the Events and Readings published will be set to the current date/time. This requires a copy be made of the Event/Readings as they are published in order to not corrupt the original data.
+
+Once the first event is published the replay function will calculate the wait time to use before sending the next Event from the recorded data. This will be based on the time difference from the original timestamp of the previous event published and the timestamp of the next event multiplied by the inverse of the `Replay Rate` specified in the request DTO. 
```

Comment:
Really like this!  I assume it is our best effort at hitting the replay rate as desired.  Given we are not deterministic and running on POSIX like threads or something, we can only attempt to replicate the rates.   Something we should mention in the docs that this is not completely deterministic and depends on the machine and what it is doing.